packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectUtil.kt

```diff
@@ -32,17 +31,21 @@ internal fun <T : RealmObject> RealmObjectInternal.manage(
     type: KClass<T>,
     objectPointer: NativePointer
 ): T {
-    val className = type.simpleName ?: sdkError("Couldn't obtain class name for $type")
-    this.`$realm$IsManaged` = true
-    this.`$realm$Owner` = realm
-    this.`$realm$Mediator` = mediator
-    this.`$realm$ObjectPointer` = objectPointer
-    this.`$realm$ClassName` = if (this is DynamicRealmObject) {
-        RealmInterop.realm_get_class(`$realm$Owner`!!.dbPointer, RealmInterop.realm_object_get_table(objectPointer)).name
-    } else {
-        realmObjectCompanionOrThrow(type).`$realm$className`
+    this.`$realm$objectReference` = ObjectReference(type).apply {
```

Comment:
Can't we move all this to the constructor of `ObjectReference`. All the properties can now be `val`s that are initialized in the constructor, right? Then we can enforce the invariants that we never operate on `ObjectReference`s without them being fully initialized!?